### ISSUE TITLE
Story information deleted when using 𠂇 character

### ISSUE THREAD

**Issue by ukanuk** [fabd/kanji-koohii#169]
I'm trying to save and share the following story on [左](https://kanji.koohii.com/study/kanji/81). When I click "save", everything looks fine. Refreshing the page, however, reveals that everything starting from the 𠂇 character is deleted. It seems like there's some incompatibility with this character, which I copied from https://en.wiktionary.org/wiki/𠂇. I've temporarily replaced the 𠂇 character with 十 in my stories, but clearly "by one's side" is not the same primitive as "ten" so it's slightly confusing. 

Before I realized what was happening, I re-typed my whole story probably 3 or 4 times, which was frustrating. Okay sure I'm dumb for not saving it elsewhere after the 1st time it happened, but still at the very least just the incompatible character should be removed, not the entire story following that one character. 

Desired story:
> For 𠂇 stroke order in {左}, {右}, {有}, and {布}, note the 1st stroke of 𠂇 follows direction of 1st stroke of inner primitive. So 𠂇 starts with downward stroke on 右有布 to match first stroke being downward on 月巾口, whereas 𠂇 starts with horizontal stroke on 左 to match first stroke going across on 工. Slight exception to this is {厷}, used in {雄}, {雄}, {紘}, {宏}, and {肱}. The 厶 inner part's first stroke is a mix of diagonal and horizontal, and the outer 𠂇 part's first stroke is horizontal (+mikemorr, +torida).

Story after saving:
> For

**Comment by iveskins** [comment]
I could reproduce this. When saving a story, all characters after that character are truncated.

**Comment by iveskins** [comment]
As a work around, instead of 𠂇 https://glyphwiki.org/wiki/u20087 try ナ https://glyphwiki.org/wiki/u30ca ?

**Comment by ukanuk** [comment]
I imagine this could be a very difficult issue to properly resolve if certain software simply doesn't support all the Unicode characterset. If that is a big task, could the site check for unsupported characters before allowing me to save my story? This would be an extension of the check it already does with {} links. For example, putting `{𠂇}` into a story currently creates this error message and won't let me save until I fix it:
**The link "{𠂇}" is not a valid Heisig or CJK Unified Ideograph character/index.**

Thank you for the suggestion to use katakana "na" instead of 十, that is a visually much more appealing workaround. I just wish my user experience hadn't been so sucky, with data appearing to be saved and not realizing the fundamental issue until I'd already wasted a lot of time. And I don't feel like I was doing anything particularly fancy in this case, I just copied it from the [Wiktionary composition of 有](https://en.wiktionary.org/wiki/有#Translingual) without hardly thinking about it. 

After a bit more testing, it appears that all of the CJK Unified Ideographs Extension B cause the same behavior. I tested characters from Wikipedia's [List of CJK Unified Ideographs Extension B (Part 1 of 7)](https://en.wikipedia.org/wiki/List_of_CJK_Unified_Ideographs_Extension_B_(Part_1_of_7)) and also its [Part 7 of 7](https://en.wikipedia.org/wiki/List_of_CJK_Unified_Ideographs_Extension_B_(Part_7_of_7)). Everything I tested caused the same behavior where the problematic character is cut out along with everything following it.

**Comment by fabd** [comment]
Thanks for the great report. Sounds like a php encoding issue. Knowing it's characters from Extension B will be very useful. I'm in the process of building a new PC, but once i have time will see if there is a solution, if not then I think it's doable to check and let user know before storing (that is, hopefully the encoding is on the database side, and not an issue when receiving the POST parameters).

**Comment by NumesSanguis** [comment]
Encountered same issue with https://en.wiktionary.org/wiki/𠂤 when I wrote a story for https://kanji.koohii.com/study/kanji/阜.
Didn't notice that the story was not properly saved and now weeks later, the story only contains a few words (everything after using 𠂤 disappeared). No idea anymore what I wrote :)

Maybe it has something to do with the CJK code length? Most Kanji have U+ 4 characters. However, Kanji like 𠂤 have 5: U+200A4 (CJK UNIFIED IDEOGRAPH-200A4)

**Comment by fabd** [comment]
@NumesSanguis thanks for the feedback. This character seems to take 4 bytes, instead of 3. Default `utf8` character set in MySQL does not store 4 byte characters.

[Stories table uses utf8 charset](https://github.com/fabd/kanji-koohii/blob/master/src/data/schemas/revtk_schema.sql#L245).

StackOverflow post suggests to use `utf8mb4`... this is a scary change... I'll have to download a most recent backup. Run the update locally which may be very slow.. make sure no data is lost.. then replace the live db. My concern here is whether this will further inflate the live database size. 

Further search suggests for VARCHAR columns it shouldn't impact too much as 3 byte characters should still take 3 bytes, and most characters in a story are within the 3-byte-utf-8 unicode blocks.

**Comment by NumesSanguis** [comment]
TL;DR: Request for **user warning** when saving the story until this issue has been solved.

Not sure how the progress is on this, but I just tried to make a story for 在 using the 𠂇, like the OP, and only realized after visiting the page at a later date that the rest of the story is gone after the use of 𠂇.

I can imagine that making changes to a database can be scary / time consuming to make sure everything is done well. However, as a temporary measurement, would it be possible to give the user feedback that the story contained character(s) that couldn't be saved in the database?

As long as you don't navigate away from the page, the story looks fine. If there is a warning, I can try to figure out which character cannot be stored and temporary do something like OP, or store it in a .txt file or something.

> I've temporarily replaced the 𠂇 character with 十 in my stories

Please warn me that I'm about to lose my "creative" story, which I won't be able to remember the next day :)

**Comment by fabd** [comment]
This is finally fixed and live. Sorry it took so long.
